lib/fs.js

```diff
@@ -229,90 +229,150 @@ fs.readFile = function(path, options, callback_) {
   var encoding = options.encoding;
   assertEncoding(encoding);
 
-  // first, stat the file, so we know the size.
-  var size;
-  var buffer; // single buffer with file data
-  var buffers; // list for when size is unknown
-  var pos = 0;
-  var fd;
-
   var flag = options.flag || 'r';
-  fs.open(path, flag, 0o666, function(er, fd_) {
-    if (er) return callback(er);
-    fd = fd_;
+  if (!nullCheck(path, callback)) return;
 
-    fs.fstat(fd, function(er, st) {
-      if (er) {
-        return fs.close(fd, function() {
-          callback(er);
-        });
-      }
+  var context = new ReadFileContext(callback, encoding);
+  var req = new FSReqWrap();
+  req.context = context;
+  req.oncomplete = readFileAfterOpen;
 
-      size = st.size;
-      if (size === 0) {
-        // the kernel lies about many files.
-        // Go ahead and try to read some bytes.
-        buffers = [];
-        return read();
-      }
+  binding.open(pathModule._makeLong(path),
+               stringToFlags(flag),
+               0o666,
+               req);
+};
 
-      if (size > kMaxLength) {
-        var err = new RangeError('File size is greater than possible Buffer: ' +
-            '0x3FFFFFFF bytes');
-        return fs.close(fd, function() {
-          callback(err);
-        });
-      }
-      buffer = new Buffer(size);
-      read();
-    });
-  });
+function ReadFileContext(callback, encoding) {
+  this.fd = undefined;
+  this.size = undefined;
+  this.callback = callback;
+  this.buffers = null;
+  this.buffer = null;
+  this.pos = 0;
+  this.encoding = encoding;
+  this.err = null;
+}
 
-  function read() {
-    if (size === 0) {
-      buffer = new Buffer(8192);
-      fs.read(fd, buffer, 0, 8192, -1, afterRead);
-    } else {
-      fs.read(fd, buffer, pos, size - pos, -1, afterRead);
-    }
+ReadFileContext.prototype.read = function() {
+  var fd = this.fd;
+  var size = this.size;
+  var buffer;
+  var offset;
+  var length;
+
+  if (size === 0) {
+    buffer = this.buffer = new Buffer(8192);
```

Comment:
I'd recommend changing this to `new SlowBuffer(8192);`. Just do a `var SlowBuffer = require('buffer').SlowBuffer;` at the top of the file.